projects/rudder-controller/project/Core/Src/controls.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * controls.c
+ *
+ *  Created on: Mar 16, 2024
+ *      Author: Jasia
+ */
+#include "controls.h"
+#include "ecompass.h"
+#include "motor.h"
+#include "encoder.h"
+
+#include <stdlib.h>
+
+
+//future: turn it into variable so SOFT can send values
+#define KI 0.075
+#define KP 0.011
+#define KD 1.022
+
+#define ADC_MAX 255
+#define ADC_MIN -255
+
+#define SMALL_ANGLE 2
+
+#define ERROR_MAX 100
+#define ERROR_MIN -100
+
+#define DELAY_TIME 100
+
+#define FULL_ROTATION 360
+#define HALF_ROTATION 180
+
+//REMOVE THIS WHEN COMs has their actual implementation
+double get_desired_heading(void){
+	return 0.0;
+}
+
+
+void main_ctrl_loop(void){
+	motor_init();
+	ecompass_init();
+	encoder_init();
+
+	while (1){
+		//add watchdog timer here
+		run_pid();
+		HAL_Delay(DELAY_TIME);
+	}
+}
+
+void run_pid(void){
+	//change to double
+	static int sum_errors = 0, prev_error = 0;
+	int P_term, I_term, D_term;
+	int error, adjusted_error, current_error;
+	int PWM_value, PWM_temp = 100; //gotta change this
```

Comment:
More descriptive name for PWM_temp is needed. 100 is a magic constant. Better to use #define

Reply:
Im not a 100% sure about what the PWM_temp value should be initially (or how to figure it out). Will change it after talking to Daniel.